workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs

```diff
@@ -146,15 +168,13 @@ private static DownloadResult Download()
                 return DownloadResult.Error;
             }
 
-            AssetDatabase.Refresh();
```

Comment:
If you don't do a `Refresh` here it's likely that the DLLs may not be imported at all until something else triggers a refresh somewhere else.

For an alternate approach for the fix, it may help to take advantage of `AssetDatabase.StartAssetEditing` and `try{...}finally{AssetDatabase.StopAssetEditing()}` - this way all of the dlls will be grouped together in one import.

I was playing around with this script for an example:

```cs
using System.IO;
using UnityEditor;
using UnityEngine;
using Directory = UnityEngine.Windows.Directory;

public class TestEditorScript : AssetPostprocessor
{
    [MenuItem("Test/Imports")]
    public static void TestBatchImports()
    {
        var testFilesPath = Path.GetFullPath(Path.Combine(Application.dataPath, "test-files"));

//        AssetDatabase.StartAssetEditing();

        Directory.CreateDirectory(testFilesPath);

        try
        {
            for (var i = 0; i < 3; ++i)
            {
                var fileName = $"file-{i}.txt";
                var fullFilePath = Path.Combine(testFilesPath, fileName);
                File.WriteAllText(fullFilePath, $"{Random.Range(0, 100)}-{Random.Range(0, 100)}");

                Debug.Log($"<color=green>Triggering an import: {fileName}</color>");

                var assetPathInAssets = Path.Combine("Assets", "test-files", fileName);
                var importer = AssetImporter.GetAtPath(assetPathInAssets);
                Debug.Log($"<color=green>Importer for {assetPathInAssets} is null before first import: {(importer == null)}</color>");

                AssetDatabase.ImportAsset(assetPathInAssets);
                Debug.Log($"<color=green>Import triggered: {fileName}</color>");

                importer = AssetImporter.GetAtPath(assetPathInAssets);
                Debug.Log($"<color=green>Importer for {assetPathInAssets} is null after first import: {(importer == null)}</color>");

                Debug.Log($"<color=green>Triggering a second import: {fileName}</color>");
                AssetDatabase.ImportAsset(assetPathInAssets);
                Debug.Log($"<color=green>Second import triggered: {fileName}</color>");
            }
        }
        finally
        {
//            AssetDatabase.StopAssetEditing();
        }
    }

    static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAssets, string[] movedAssets,
        string[] movedFromAssetPaths)
    {
        if (importedAssets.Length > 0)
        {
            Debug.Log($"<color=red>Imported assets: {string.Join(",", importedAssets)}</color>");
        }

        if (deletedAssets.Length > 0)
        {
            Debug.Log($"<color=red>Deleted assets: {string.Join(",", deletedAssets)}</color>");
        }
    }
}
```

In the default configuration the script creates a few files and imports them one by one (without using `AssetDatabase.StartAssetEditing()`).

If you copy in this script to a Unity project and select "Test/Imports", you get this sequence of logs:
![image](https://user-images.githubusercontent.com/31209570/46567357-c4f47f80-c929-11e8-8cb3-3eb4b386939e.png)

This highlights a few things (some of these things we may know already but may help to highlight):
- the asset importer is null before the first import for the asset happens (i.e. before the AssetPostprocessor gets executed)
- the imports are synchronous, so one import will finish before `ImportAsset` returns
- after the import is completed, the asset importer is not null any more, so we can now modify import settings

For the next step:
- delete the `test-files` folder, 
- uncomment `StartAssetEditing` and `StopAssetEditing` to enable batch imports 
- select **Test** / **Imports**

![image](https://user-images.githubusercontent.com/31209570/46567368-e05f8a80-c929-11e8-8bcc-6ff4c5b3f78f.png)

This run highlights:
- First of all everything is much faster! (can see this from the timestamps too)
- However, asset importers are null even after the first "import" call is finished (probably because the "import" call now only flags for `StopAssetEditing` to do the importing) - so guessing they will not be null after `StopAssetEditing`

To test the last guess:
```cs

        finally
        {
            AssetDatabase.StopAssetEditing();
            
            for (var i = 0; i < 3; ++i)
            {
                var fileName = $"file-{i}.txt";

                var assetPathInAssets = Path.Combine("Assets", "test-files", fileName);
                var importer = AssetImporter.GetAtPath(assetPathInAssets);
                Debug.Log($"<color=green>Importer for {assetPathInAssets} is null after StopAssetEditing: {(importer == null)}</color>");
            }
        }
```

![image](https://user-images.githubusercontent.com/31209570/46567833-6c75b000-c932-11e8-842f-34041113cbf5.png)

--------------

## Recommendation

Based on these findings, my recommendation would be to do this:

- `private static DownloadResult Download()`
  - StartAssetEditing()
  - Download all dlls
  - Trigger `AssetDatabase.ImportAsset` on all written dll paths within assets
  - StopAssetEditing() (let dlls get imported first time, likely with wrong settings!)
  - SetPluginsCompatibility()
- `private static void SetPluginsCompatibility()`
  - StartAssetEditing
  - Iterate over all written dll paths within assets:
    - Adjust importer settings if necessary (we know _(†)_ importers have to exist at this point)
    - Trigger a reimport if necessary
  - StopAssetEditing (let dlls get imported with correct importer settings)

I think it would not need the post processor in this case, unless we want to handle the situation when the user (or something else) replaces the plugin dlls outside these methods.

-----------

## † Scary story time

Note: **I'm not sure** what happens during `StopAssetEditing` if one import fails (for example, if two dlls have same name and same platform). If the importing process for DLLs does not continue for all DLLs in this case, at worst case multiple passes of SetPluginsCompatibility may need to happen, until AssetImporters exist and have been configured correctly for all DLLs.

This is probably unlikely but still may be worth a test.

Could the next few paragraphs happen for example?

In near worst case, we have let's say 5 very unfortunate dlls.
- first one gets imported with default settings (wrong)
- second one tries to get imported with default settings (wrong), fails, process stops here?
- third one does not get imported
- fourth one does not get imported
- fifth one does not get imported

In this case the first and second ones will have asset importers, but not the third and the fourth and the fifth. So the plugin compatibility method would be able to fix the first two, then, in near worst case (again)
- first one gets imported fine
- second one gets imported fine
- third one gets imported with default settings (wrong) but does not fail because it may only clash with the fourth (so evil)
- fourth one tries to get imported with default settings (wrong), fails
- fifth one does not get imported

Now first, second are all good, third and fourth can be fixed because there will be importers, but fifth is still in trouble and without an importer to fix.

Plugin compatibility needs to run again, new result:
- first one gets imported fine
- second one gets imported fine
- third one gets imported fine
- fourth one gets imported fine
- fifth one gets imported with default settings (wrong)

Plugin compatibility needs to run again, it sees that all plugins do have the right import settings, so it can emit a happy beep after the final reimport.
- first one gets imported fine
- second one gets imported fine
- third one gets imported fine
- fourth one gets imported fine
- fifth one gets imported fine

Let's hope this is not necessary...

Reply:
Also have we played with different parameters to `Refresh`, e.g. `AssetDatabase.Refresh(ImportAssetOptions.ForceSynchronousImport | ImportAssetOptions.ForceUpdate | ImportAssetOptions.ImportRecursive)` ? It's possible that the default options may not have created the correct settings in time for the call to the compatibility fixer

Reply:
Thanks for the detailed research, @firtina-improbable ! Let me try and refactor the diff using your suggestion.
Re AssetDatabase.Refresh - yes, this was the first approach I've tried, unfortunately even with all options you've mentioned, AssetDatabase didn't seem to be in the complete state after the import, possibly because of the issue you've described

Reply:
Chatted offline about this - for the system to work properly we'll need to set up communication between downloading module and plugin processor. While this is something we'll need to add eventually if we are to download plugins depending on a platform, it is probably overkill for a simple issue like this one.